org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/ClasspathUpdateHandler.java

```diff
@@ -41,53 +43,67 @@
-
 	private final JavaClientConnection connection;
+	private final BaseDocumentLifeCycleHandler lifeCycleHandler;
 
 	public ClasspathUpdateHandler(JavaClientConnection client) {
+		this(client, null);
+	}
+
+	public ClasspathUpdateHandler(JavaClientConnection client, BaseDocumentLifeCycleHandler lifeCycleHandler) {
 		this.connection = client;
+		this.lifeCycleHandler = lifeCycleHandler;
 	}
 
 	@Override
 	public void elementChanged(ElementChangedEvent event) {
 		// Collect project names which have classpath changed.
-		Set<String> uris = processDelta(event.getDelta(), null);
-		if (connection != null && uris != null && !uris.isEmpty()) {
-			for (String uri : uris) {
+		Set<IJavaProject> projects = new HashSet<>();
+		processDelta(event.getDelta(), projects);
+		if (connection != null && projects != null && !projects.isEmpty()) {
+			for (IJavaProject javaProject : projects) {
+				String uri = ProjectUtils.getProjectRealFolder(javaProject.getProject()).toFile().toURI().toString();
 				PreferenceManager preferenceManager = JavaLanguageServerPlugin.getPreferencesManager();
-				if (preferenceManager.getPreferences().getNullAnalysisMode().equals(FeatureStatus.automatic)) {
-					IProject project = ProjectUtils.getProjectFromUri(uri);
-					if (project != null) {
-						IJavaProject javaProject = ProjectUtils.getJavaProject(project);
-						if (javaProject != null) {
-							WorkspaceJob job = new WorkspaceJob("Classpath Update Job") {
-								@Override
-								public IStatus runInWorkspace(IProgressMonitor monitor) {
-									if (!preferenceManager.getPreferences().updateAnnotationNullAnalysisOptions(javaProject, true) || !preferenceManager.getPreferences().isAutobuildEnabled()) {
-										// When the project's compiler options didn't change or auto build is disabled, rebuilding is not required.
-										return Status.OK_STATUS;
-									}
-									BuildWorkspaceHandler buildWorkspaceHandler = new BuildWorkspaceHandler(JavaLanguageServerPlugin.getProjectsManager());
-									ProjectBuildParams params = new ProjectBuildParams(Arrays.asList(new TextDocumentIdentifier(uri)), true);
-									BuildWorkspaceStatus status = buildWorkspaceHandler.buildProjects(params, monitor);
-									switch (status) {
-										case FAILED:
-										case WITH_ERROR:
-											return Status.error("error occurs during building project");
-										case SUCCEED:
-											return Status.OK_STATUS;
-										case CANCELLED:
-											return Status.CANCEL_STATUS;
-										default:
-											return Status.OK_STATUS;
-									}
+				if (!preferenceManager.getPreferences().isAutobuildEnabled()) {
+					if (lifeCycleHandler != null) {
+						for (ICompilationUnit unit : JavaCore.getWorkingCopies(null)) {
+							if (unit.getJavaProject().equals(javaProject)) {
+								try {
+									lifeCycleHandler.triggerValidation(unit);
+									throw new JavaModelException(new CoreException(null));
```

Comment:
I assume this was from testing the logging after a failure of triggering validation. It should be removed.

Reply:
Oops... I removed it.